File: zig/src/indicators/common/triangular_moving_average/triangular_moving_average.zig
const std = @import("std");
const math = std.math;


const entities = @import("entities");
const Bar = entities.Bar;
const Quote = entities.Quote;
const Trade = entities.Trade;
const Scalar = entities.Scalar;
const bar_component = entities.bar_component;
const quote_component = entities.quote_component;
const trade_component = entities.trade_component;
const indicator_mod = @import("../../core/indicator.zig");
const line_indicator_mod = @import("../../core/line_indicator.zig");
const build_metadata_mod = @import("../../core/build_metadata.zig");
const component_triple_mnemonic_mod = @import("../../core/component_triple_mnemonic.zig");
const identifier_mod = @import("../../core/identifier.zig");
const metadata_mod = @import("../../core/metadata.zig");

const OutputArray = indicator_mod.OutputArray;
const LineIndicator = line_indicator_mod.LineIndicator;
const Identifier = identifier_mod.Identifier;
const Metadata = metadata_mod.Metadata;

/// Enumerates the outputs of the triangular moving average indicator.
pub const TriangularMovingAverageOutput = enum(u8) {
    value = 1,
};

/// Parameters to create an instance of the triangular moving average indicator.
pub const TriangularMovingAverageParams = struct {
    length: usize,
    bar_component: ?bar_component.BarComponent = null,
    quote_component: ?quote_component.QuoteComponent = null,
    trade_component: ?trade_component.TradeComponent = null,
};

/// Computes the triangular moving average (TRIMA).
///
/// The TRIMA puts more weight on the data in the middle of the window,
/// equivalent to doing a SMA of a SMA.
///
/// The indicator is not primed during the first ℓ−1 updates.
pub const TriangularMovingAverage = struct {
    line: LineIndicator,
    factor: f64,
    numerator: f64,
    numerator_sub: f64,
    numerator_add: f64,
    window: []f64,
    window_length: usize,
    window_length_half: usize,
    window_count: usize,
    is_odd: bool,
    primed: bool,
    allocator: std.mem.Allocator,
    mnemonic_buf: [64]u8,
    mnemonic_len: usize,
    description_buf: [128]u8,
    description_len: usize,

    pub fn init(allocator: std.mem.Allocator, params: TriangularMovingAverageParams) !TriangularMovingAverage {
        if (params.length < 2) {
            return error.InvalidLength;
        }

        const bc = params.bar_component orelse bar_component.default_bar_component;
        const qc = params.quote_component orelse quote_component.default_quote_component;
        const tc = params.trade_component orelse trade_component.default_trade_component;

        var triple_buf: [64]u8 = undefined;
        const triple = component_triple_mnemonic_mod.componentTripleMnemonic(&triple_buf, bc, qc, tc);

        var mnemonic_buf: [64]u8 = undefined;
        const mnemonic_slice = std.fmt.bufPrint(&mnemonic_buf, "trima({d}{s})", .{ params.length, triple }) catch
            return error.MnemonicTooLong;
        const mnemonic_len = mnemonic_slice.len;

        var description_buf: [128]u8 = undefined;
        const desc_slice = std.fmt.bufPrint(&description_buf, "Triangular moving average {s}", .{mnemonic_slice}) catch
            return error.MnemonicTooLong;
        const description_len = desc_slice.len;

        var length_half = params.length >> 1;
        const l = 1 + length_half;
        const is_odd = params.length % 2 == 1;

        var factor: f64 = undefined;
        if (is_odd) {
            factor = 1.0 / @as(f64, @floatFromInt(l * l));
        } else {
            factor = 1.0 / @as(f64, @floatFromInt(length_half * l));
            length_half -= 1;
        }

        const window = try allocator.alloc(f64, params.length);
        @memset(window, 0.0);

        return .{
            .line = LineIndicator.new(
                mnemonic_buf[0..mnemonic_len],
                description_buf[0..description_len],
                params.bar_component,
                params.quote_component,
                params.trade_component,
            ),
            .factor = factor,
            .numerator = 0.0,
            .numerator_sub = 0.0,
            .numerator_add = 0.0,
            .window = window,
            .window_length = params.length,
            .window_length_half = length_half,
            .window_count = 0,
            .is_odd = is_odd,
            .primed = false,
            .allocator = allocator,
            .mnemonic_buf = mnemonic_buf,
            .mnemonic_len = mnemonic_len,
            .description_buf = description_buf,
            .description_len = description_len,
        };
    }

    pub fn deinit(self: *TriangularMovingAverage) void {
        self.allocator.free(self.window);
    }

    pub fn fixSlices(self: *TriangularMovingAverage) void {
        self.line.mnemonic = self.mnemonic_buf[0..self.mnemonic_len];
        self.line.description = self.description_buf[0..self.description_len];
    }

    pub fn update(self: *TriangularMovingAverage, sample: f64) f64 {
        if (math.isNan(sample)) {
            return sample;
        }

        var temp = sample;

        if (self.primed) {
            self.numerator -= self.numerator_sub;
            self.numerator_sub -= self.window[0];

            const j = self.window_length - 1;
            var i: usize = 0;
            while (i < j) : (i += 1) {
                self.window[i] = self.window[i + 1];
            }

            self.window[j] = temp;
            temp = self.window[self.window_length_half];
            self.numerator_sub += temp;

            if (self.is_odd) {
                self.numerator += self.numerator_add;
                self.numerator_add -= temp;
            } else {
                self.numerator_add -= temp;
                self.numerator += self.numerator_add;
            }

            temp = sample;
            self.numerator_add += temp;
            self.numerator += temp;
        } else {
            self.window[self.window_count] = temp;
            self.window_count += 1;

            if (self.window_length > self.window_count) {
                return math.nan(f64);
            }

            // Priming: compute initial numerator_sub and numerator
            {
                var ii: usize = self.window_length_half + 1;
                while (ii > 0) {
                    ii -= 1;
                    self.numerator_sub += self.window[ii];
                    self.numerator += self.numerator_sub;
                }
            }

            {
                var ii: usize = self.window_length_half + 1;
                while (ii < self.window_length) : (ii += 1) {
                    self.numerator_add += self.window[ii];
                    self.numerator += self.numerator_add;
                }
            }

            self.primed = true;
        }

        return self.numerator * self.factor;
    }

    pub fn isPrimed(self: *const TriangularMovingAverage) bool {
        return self.primed;
    }

    pub fn getMetadata(self: *const TriangularMovingAverage, out: *Metadata) void {
        build_metadata_mod.buildMetadata(
            out,
            .triangular_moving_average,
            self.line.mnemonic,
            self.line.description,
            &[_]build_metadata_mod.OutputText{
                .{ .mnemonic = self.line.mnemonic, .description = self.line.description },
            },
        );
    }

    pub fn updateScalar(self: *TriangularMovingAverage, sample: *const Scalar) OutputArray {
        const value = self.update(sample.value);
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateBar(self: *TriangularMovingAverage, sample: *const Bar) OutputArray {
        const value = self.update(self.line.extractBar(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateQuote(self: *TriangularMovingAverage, sample: *const Quote) OutputArray {
        const value = self.update(self.line.extractQuote(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateTrade(self: *TriangularMovingAverage, sample: *const Trade) OutputArray {
        const value = self.update(self.line.extractTrade(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn indicator(self: *TriangularMovingAverage) indicator_mod.Indicator {
        return .{
            .ptr = @ptrCast(self),
            .vtable = &vtable,
        };
    }

    const vtable = indicator_mod.Indicator.VTable{
        .isPrimed = vtableIsPrimed,
        .metadata = vtableMetadata,
        .updateScalar = vtableUpdateScalar,
        .updateBar = vtableUpdateBar,
        .updateQuote = vtableUpdateQuote,
        .updateTrade = vtableUpdateTrade,
    };

    fn vtableIsPrimed(ptr: *anyopaque) bool {
        const self: *TriangularMovingAverage = @ptrCast(@alignCast(ptr));
        return self.isPrimed();
    }

    fn vtableMetadata(ptr: *anyopaque, out: *Metadata) void {
        const self: *const TriangularMovingAverage = @ptrCast(@alignCast(ptr));
        self.getMetadata(out);
    }

    fn vtableUpdateScalar(ptr: *anyopaque, sample: *const Scalar) OutputArray {
        const self: *TriangularMovingAverage = @ptrCast(@alignCast(ptr));
        return self.updateScalar(sample);
    }

    fn vtableUpdateBar(ptr: *anyopaque, sample: *const Bar) OutputArray {
        const self: *TriangularMovingAverage = @ptrCast(@alignCast(ptr));
        return self.updateBar(sample);
    }

    fn vtableUpdateQuote(ptr: *anyopaque, sample: *const Quote) OutputArray {
        const self: *TriangularMovingAverage = @ptrCast(@alignCast(ptr));
        return self.updateQuote(sample);
    }

    fn vtableUpdateTrade(ptr: *anyopaque, sample: *const Trade) OutputArray {
        const self: *TriangularMovingAverage = @ptrCast(@alignCast(ptr));
        return self.updateTrade(sample);
    }

    pub const InitError = error{
        InvalidLength,
        MnemonicTooLong,
        OutOfMemory,
    };
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

const testing = std.testing;
const testdata = @import("testdata.zig");


fn createTrima(allocator: std.mem.Allocator, length: usize) !TriangularMovingAverage {
    var trima = try TriangularMovingAverage.init(allocator, .{ .length = length });
    trima.fixSlices();
    return trima;
}

fn runAll(trima: *TriangularMovingAverage, input: *const [252]f64) [252]f64 {
    var results: [252]f64 = undefined;
    for (0..252) |i| {
        results[i] = trima.update(input[i]);
    }
    return results;
}

test "trima update length 9" {
    const input = testdata.testInput();
    var trima = try createTrima(testing.allocator, 9);
    defer trima.deinit();
    const results = runAll(&trima, &input);

    for (0..8) |i| {
        try testing.expect(math.isNan(results[i]));
    }
    try testing.expect(@abs(93.8176 - results[8]) < 1e-4);
    try testing.expect(@abs(109.1312 - results[251]) < 1e-4);
    try testing.expect(math.isNan(trima.update(math.nan(f64))));
}

test "trima update length 10" {
    const input = testdata.testInput();
    var trima = try createTrima(testing.allocator, 10);
    defer trima.deinit();
    const results = runAll(&trima, &input);

    for (0..9) |i| {
        try testing.expect(math.isNan(results[i]));
    }
    try testing.expect(@abs(93.6043 - results[9]) < 1e-4);
    try testing.expect(@abs(93.4252 - results[10]) < 1e-4);
    try testing.expect(@abs(109.1850 - results[250]) < 1e-4);
    try testing.expect(@abs(109.1407 - results[251]) < 1e-4);
    try testing.expect(math.isNan(trima.update(math.nan(f64))));
}

test "trima update length 12" {
    const input = testdata.testInput();
    var trima = try createTrima(testing.allocator, 12);
    defer trima.deinit();
    const results = runAll(&trima, &input);

    for (0..10) |i| {
        try testing.expect(math.isNan(results[i]));
    }
    try testing.expect(@abs(93.5329 - results[11]) < 1e-4);
    try testing.expect(@abs(109.1157 - results[251]) < 1e-4);
    try testing.expect(math.isNan(trima.update(math.nan(f64))));
}

test "trima is primed length 9" {
    const input = testdata.testInput();
    var trima = try createTrima(testing.allocator, 9);
    defer trima.deinit();

    try testing.expect(!trima.isPrimed());
    for (0..8) |i| {
        _ = trima.update(input[i]);
        try testing.expect(!trima.isPrimed());
    }
    _ = trima.update(input[8]);
    try testing.expect(trima.isPrimed());
}

test "trima is primed length 12" {
    const input = testdata.testInput();
    var trima = try createTrima(testing.allocator, 12);
    defer trima.deinit();

    try testing.expect(!trima.isPrimed());
    for (0..11) |i| {
        _ = trima.update(input[i]);
        try testing.expect(!trima.isPrimed());
    }
    _ = trima.update(input[11]);
    try testing.expect(trima.isPrimed());
}

test "trima metadata" {
    var trima = try createTrima(testing.allocator, 5);
    defer trima.deinit();
    var m: Metadata = undefined;
    trima.getMetadata(&m);

    try testing.expectEqual(Identifier.triangular_moving_average, m.identifier);
    try testing.expectEqual(@as(usize, 1), m.outputs_len);
    try testing.expectEqual(@as(i32, 1), m.outputs_buf[0].kind);
    try testing.expectEqualStrings("trima(5)", m.outputs_buf[0].mnemonic);
    try testing.expectEqualStrings("Triangular moving average trima(5)", m.outputs_buf[0].description);
}

test "trima init invalid length" {
    const result = TriangularMovingAverage.init(testing.allocator, .{ .length = 1 });
    try testing.expectError(error.InvalidLength, result);

    const result0 = TriangularMovingAverage.init(testing.allocator, .{ .length = 0 });
    try testing.expectError(error.InvalidLength, result0);
}

test "trima update entity" {
    const input = testdata.testInput();
    const time: i64 = 1617235200;
    const exp: f64 = 93.5329761904762;

    // scalar
    {
        var trima = try createTrima(testing.allocator, 12);
        defer trima.deinit();
        for (0..11) |i| {
            _ = trima.update(input[i]);
        }
        const out = trima.updateScalar(&.{ .time = time, .value = input[11] });
        try testing.expectEqual(@as(usize, 1), out.len);
        const s = out.slice()[0].scalar;
        try testing.expectEqual(time, s.time);
        try testing.expect(@abs(exp - s.value) < 1e-12);
    }

    // bar
    {
        var trima = try createTrima(testing.allocator, 12);
        defer trima.deinit();
        for (0..11) |i| {
            _ = trima.update(input[i]);
        }
        const bar = Bar{ .time = time, .open = 0, .high = 0, .low = 0, .close = input[11], .volume = 0 };
        const out = trima.updateBar(&bar);
        const s = out.slice()[0].scalar;
        try testing.expect(@abs(exp - s.value) < 1e-12);
    }
}

test "trima mnemonic components" {
    {
        var trima = try createTrima(testing.allocator, 5);
        defer trima.deinit();
        try testing.expectEqualStrings("trima(5)", trima.line.mnemonic);
    }
    {
        var trima = try TriangularMovingAverage.init(testing.allocator, .{
            .length = 5,
            .bar_component = .median,
        });
        defer trima.deinit();
        trima.fixSlices();
        try testing.expectEqualStrings("trima(5, hl/2)", trima.line.mnemonic);
    }
}

File: rs/src/indicators/common/triangular_moving_average/triangular_moving_average.rs
use crate::entities::bar::Bar;
use crate::entities::bar_component::{component_value as bar_component_value, BarComponent, DEFAULT_BAR_COMPONENT};
use crate::entities::quote::Quote;
use crate::entities::quote_component::{component_value as quote_component_value, QuoteComponent, DEFAULT_QUOTE_COMPONENT};
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::entities::trade_component::{component_value as trade_component_value, TradeComponent, DEFAULT_TRADE_COMPONENT};
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::component_triple_mnemonic::component_triple_mnemonic;
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::{Indicator, Output};
use crate::indicators::core::line_indicator::LineIndicator;
use crate::indicators::core::metadata::Metadata;

// ---------------------------------------------------------------------------
// Params
// ---------------------------------------------------------------------------

/// Parameters to create an instance of the triangular moving average indicator.
pub struct TriangularMovingAverageParams {
    /// The length (number of time periods) of the moving window.
    /// Must be greater than 1.
    pub length: usize,
    /// Bar component to extract. `None` means use default (Close).
    pub bar_component: Option<BarComponent>,
    /// Quote component to extract. `None` means use default (Mid).
    pub quote_component: Option<QuoteComponent>,
    /// Trade component to extract. `None` means use default (Price).
    pub trade_component: Option<TradeComponent>,
}

impl Default for TriangularMovingAverageParams {
    fn default() -> Self {
        Self {
            length: 20,
            bar_component: None,
            quote_component: None,
            trade_component: None,
        }
    }
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

/// Enumerates the outputs of the triangular moving average indicator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum TriangularMovingAverageOutput {
    /// The scalar value of the moving average.
    Value = 1,
}

// ---------------------------------------------------------------------------
// Indicator
// ---------------------------------------------------------------------------

/// Computes the triangular moving average (TRIMA).
///
/// The TRIMA puts more weight on the data in the middle of the window,
/// equivalent to computing an SMA of an SMA.
///
/// Uses an optimised incremental algorithm with four adjustments per step.
pub struct TriangularMovingAverage {
    line: LineIndicator,
    factor: f64,
    numerator: f64,
    numerator_sub: f64,
    numerator_add: f64,
    window: Vec<f64>,
    window_length: usize,
    window_length_half: usize,
    window_count: usize,
    is_odd: bool,
    primed: bool,
}

impl TriangularMovingAverage {
    /// Creates a new TriangularMovingAverage from the given parameters.
    pub fn new(params: &TriangularMovingAverageParams) -> Result<Self, String> {
        if params.length < 2 {
            return Err("invalid triangular moving average parameters: length should be greater than 1".to_string());
        }

        let bc = params.bar_component.unwrap_or(DEFAULT_BAR_COMPONENT);
        let qc = params.quote_component.unwrap_or(DEFAULT_QUOTE_COMPONENT);
        let tc = params.trade_component.unwrap_or(DEFAULT_TRADE_COMPONENT);

        let bar_func = bar_component_value(bc);
        let quote_func = quote_component_value(qc);
        let trade_func = trade_component_value(tc);

        let length = params.length;
        let length_half = length >> 1;
        let l = 1 + length_half;
        let is_odd = length % 2 == 1;

        let (factor, window_length_half) = if is_odd {
            // Odd: 1+2+...+(l)+...+2+1 = l*l where l = (length+1)/2 = length_half+1.
            (1.0 / (l * l) as f64, length_half)
        } else {
            // Even: 1+2+...+l+l+...+2+1 = length_half * l.
            (1.0 / (length_half * l) as f64, length_half - 1)
        };

        let mnemonic = format!("trima({}{})", length, component_triple_mnemonic(bc, qc, tc));
        let description = format!("Triangular moving average {}", mnemonic);

        let line = LineIndicator::new(mnemonic, description, bar_func, quote_func, trade_func);

        Ok(Self {
            line,
            factor,
            numerator: 0.0,
            numerator_sub: 0.0,
            numerator_add: 0.0,
            window: vec![0.0; length],
            window_length: length,
            window_length_half,
            window_count: 0,
            is_odd,
            primed: false,
        })
    }

    /// Core update logic. Returns the TRIMA value or NaN if not yet primed.
    pub fn update(&mut self, sample: f64) -> f64 {
        if sample.is_nan() {
            return sample;
        }

        let temp = sample;

        if self.primed {
            self.numerator -= self.numerator_sub;
            self.numerator_sub -= self.window[0];

            let j = self.window_length - 1;
            for i in 0..j {
                self.window[i] = self.window[i + 1];
            }

            self.window[j] = temp;
            let mid = self.window[self.window_length_half];
            self.numerator_sub += mid;

            if self.is_odd {
                self.numerator += self.numerator_add;
                self.numerator_add -= mid;
            } else {
                self.numerator_add -= mid;
                self.numerator += self.numerator_add;
            }

            self.numerator_add += sample;
            self.numerator += sample;
        } else {
            self.window[self.window_count] = temp;
            self.window_count += 1;

            if self.window_length > self.window_count {
                return f64::NAN;
            }

            // Initialise numerator_sub from the middle going left.
            let half = self.window_length_half;
            for i in (0..=half).rev() {
                self.numerator_sub += self.window[i];
                self.numerator += self.numerator_sub;
            }

            // Initialise numerator_add from the middle+1 going right.
            for i in (half + 1)..self.window_length {
                self.numerator_add += self.window[i];
                self.numerator += self.numerator_add;
            }

            self.primed = true;
        }

        self.numerator * self.factor
    }
}

impl Indicator for TriangularMovingAverage {
    fn is_primed(&self) -> bool {
        self.primed
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::TriangularMovingAverage,
            &self.line.mnemonic,
            &self.line.description,
            &[OutputText {
                mnemonic: self.line.mnemonic.clone(),
                description: self.line.description.clone(),
            }],
        )
    }

    fn update_scalar(&mut self, sample: &Scalar) -> Output {
        let value = self.update(sample.value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_bar(&mut self, sample: &Bar) -> Output {
        let sample_value = (self.line.bar_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_quote(&mut self, sample: &Quote) -> Output {
        let sample_value = (self.line.quote_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_trade(&mut self, sample: &Trade) -> Output {
        let sample_value = (self.line.trade_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::testdata::testdata;
    use crate::entities::bar_component::BarComponent;
    use crate::indicators::core::outputs::shape::Shape;
    fn create_trima(length: usize) -> TriangularMovingAverage {
        TriangularMovingAverage::new(&TriangularMovingAverageParams { length, ..Default::default() }).unwrap()
    }

    #[test]
    fn test_update_length_9() {
        let mut trima = create_trima(9);
        let input = testdata::test_input();

        for i in 0..8 {
            assert!(trima.update(input[i]).is_nan(), "[{}] expected NaN", i);
        }

        let act = trima.update(input[8]);
        assert!((93.8176 - act).abs() < 1e-4, "[8] expected 93.8176, got {}", act);

        for i in 9..input.len() - 1 {
            trima.update(input[i]);
        }

        let act = trima.update(input[251]);
        assert!((109.1312 - act).abs() < 1e-4, "[251] expected 109.1312, got {}", act);

        assert!(trima.update(f64::NAN).is_nan());
    }

    #[test]
    fn test_update_length_10() {
        let mut trima = create_trima(10);
        let input = testdata::test_input();

        for i in 0..9 {
            assert!(trima.update(input[i]).is_nan(), "[{}] expected NaN", i);
        }

        let act = trima.update(input[9]);
        assert!((93.6043 - act).abs() < 1e-4, "[9] expected 93.6043, got {}", act);

        let act = trima.update(input[10]);
        assert!((93.4252 - act).abs() < 1e-4, "[10] expected 93.4252, got {}", act);

        for i in 11..250 {
            trima.update(input[i]);
        }

        let act = trima.update(input[250]);
        assert!((109.1850 - act).abs() < 1e-4, "[250] expected 109.1850, got {}", act);

        let act = trima.update(input[251]);
        assert!((109.1407 - act).abs() < 1e-4, "[251] expected 109.1407, got {}", act);

        assert!(trima.update(f64::NAN).is_nan());
    }

    #[test]
    fn test_update_length_12() {
        let mut trima = create_trima(12);
        let input = testdata::test_input();

        for i in 0..10 {
            assert!(trima.update(input[i]).is_nan(), "[{}] expected NaN", i);
        }

        // index 10 is still NaN for length 12 (need 11 NaN values: indices 0..=10)
        assert!(trima.update(input[10]).is_nan(), "[10] expected NaN");

        let act = trima.update(input[11]);
        assert!((93.5329 - act).abs() < 1e-4, "[11] expected 93.5329, got {}", act);

        for i in 12..251 {
            trima.update(input[i]);
        }

        let act = trima.update(input[251]);
        assert!((109.1157 - act).abs() < 1e-4, "[251] expected 109.1157, got {}", act);

        assert!(trima.update(f64::NAN).is_nan());
    }

    #[test]
    fn test_update_xls12() {
        let mut trima = create_trima(12);
        let input = testdata::test_input();
        let expected = testdata::expected_xls12();

        for i in 0..11 {
            assert!(trima.update(input[i]).is_nan(), "[{}] expected NaN", i);
        }

        for i in 11..input.len() {
            let act = trima.update(input[i]);
            assert!(
                (expected[i] - act).abs() < 1e-12,
                "[{}] expected {}, got {}", i, expected[i], act
            );
        }
    }

    #[test]
    fn test_update_entity() {
        let length = 12;
        let input = testdata::test_input();
        let inp = 97.250000; // input[11]
        let exp = 93.5329761904762;
        let time = 1617235200_i64;

        // scalar
        let mut trima = create_trima(length);
        for i in 0..11 {
            trima.update(input[i]);
        }
        let out = trima.update_scalar(&Scalar::new(time, inp));
        assert_eq!(out.len(), 1);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert_eq!(s.time, time);
        assert!((exp - s.value).abs() < 1e-12);

        // bar
        let mut trima = create_trima(length);
        for i in 0..11 {
            trima.update(input[i]);
        }
        let bar = Bar::new(time, 0.0, 0.0, 0.0, inp, 0.0);
        let out = trima.update_bar(&bar);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert!((exp - s.value).abs() < 1e-12);

        // quote
        let mut trima = create_trima(length);
        for i in 0..11 {
            trima.update(input[i]);
        }
        let quote = Quote::new(time, inp, inp, 0.0, 0.0);
        let out = trima.update_quote(&quote);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert!((exp - s.value).abs() < 1e-12);

        // trade
        let mut trima = create_trima(length);
        for i in 0..11 {
            trima.update(input[i]);
        }
        let trade = Trade::new(time, inp, 0.0);
        let out = trima.update_trade(&trade);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert!((exp - s.value).abs() < 1e-12);
    }

    #[test]
    fn test_is_primed() {
        let input = testdata::test_input();

        // length 9: primed after 9 samples (index 8)
        let mut trima = create_trima(9);
        assert!(!trima.is_primed());
        for i in 0..8 {
            trima.update(input[i]);
            assert!(!trima.is_primed(), "[{}] should not be primed", i);
        }
        for i in 8..input.len() {
            trima.update(input[i]);
            assert!(trima.is_primed(), "[{}] should be primed", i);
        }

        // length 12: primed after 12 samples (index 11)
        let mut trima = create_trima(12);
        assert!(!trima.is_primed());
        for i in 0..11 {
            trima.update(input[i]);
            assert!(!trima.is_primed(), "[{}] should not be primed", i);
        }
        for i in 11..input.len() {
            trima.update(input[i]);
            assert!(trima.is_primed(), "[{}] should be primed", i);
        }
    }

    #[test]
    fn test_metadata() {
        let trima = create_trima(5);
        let m = trima.metadata();
        assert_eq!(m.identifier, Identifier::TriangularMovingAverage);
        assert_eq!(m.outputs.len(), 1);
        assert_eq!(m.outputs[0].kind, TriangularMovingAverageOutput::Value as i32);
        assert_eq!(m.outputs[0].shape, Shape::Scalar);
        assert_eq!(m.outputs[0].mnemonic, "trima(5)");
        assert_eq!(m.outputs[0].description, "Triangular moving average trima(5)");
    }

    #[test]
    fn test_new_invalid_length() {
        let r = TriangularMovingAverage::new(&TriangularMovingAverageParams { length: 1, ..Default::default() });
        assert!(r.is_err());
        assert_eq!(r.err().unwrap(), "invalid triangular moving average parameters: length should be greater than 1");

        let r = TriangularMovingAverage::new(&TriangularMovingAverageParams { length: 0, ..Default::default() });
        assert!(r.is_err());
    }

    #[test]
    fn test_mnemonic_components() {
        // all defaults -> no component suffix
        let trima = create_trima(5);
        assert_eq!(trima.line.mnemonic, "trima(5)");

        // bar component set
        let trima = TriangularMovingAverage::new(&TriangularMovingAverageParams {
            length: 5, bar_component: Some(BarComponent::Median), ..Default::default()
        }).unwrap();
        assert_eq!(trima.line.mnemonic, "trima(5, hl/2)");

        // bar and trade set
        let trima = TriangularMovingAverage::new(&TriangularMovingAverageParams {
            length: 5,
            bar_component: Some(BarComponent::High),
            quote_component: None,
            trade_component: Some(TradeComponent::Volume),
        }).unwrap();
        assert_eq!(trima.line.mnemonic, "trima(5, h, v)");
    }
}
